File: dev/hmget-tests.py
#!/usr/bin/env python3
"""Ordered nullable Hash projections across the store, twin and live hosts."""
import importlib.util
import json
from pathlib import Path
import subprocess
import sys
import tempfile

ROOT = Path(__file__).resolve().parents[1]
spec = importlib.util.spec_from_file_location('hmget_support', ROOT / 'dev/lists-tests.py')
support = importlib.util.module_from_spec(spec)
spec.loader.exec_module(support)
run, require, tet = support.run, support.require, support.tet
TAG, KEY, WRONG = support.TAG, support.KEY, support.WRONG
QUERY = {
    'lookup': [b'missing', b'z', b'a', b'absent', b'z', b''],
    'one': [b'z'],
    'repeat': [b'z', b'z', b'z'],
    'many': [f'f{i:03}'.encode() for i in range(129)],
    'binary': [support.BINARY, b'', support.BINARY, b'none'],
    'head': [b'missing', b'z'],
}
QUERY['raw'] = QUERY['within'] = QUERY['computed'] = QUERY['lookup']
ENTRIES = tuple(QUERY) + ('earlier',)
READONLY = {name + 'Script' for name in QUERY if name != 'within'}
EXAMPLE = b'["member",null,"Alice","member"]\n'


def arguments(fields):
    require(bool(fields), 'HMGET nonempty test field list')
    result = f'(bulkOne {tet(fields[-1])})'
    for field in reversed(fields[:-1]):
        result = f'(bulkMore {tet(field)} {result})'
    return result


def command(fields):
    return f'hmget Reply {TAG} (jobs b"mail") {arguments(fields)}'


def cases():
    inputs = [
        ('lookup', None), ('lookup', {b'z': b'Z', b'a': b'A', b'': b''}),
        ('lookup', {b'z': b'"\n$(touch forbidden)\\', b'a': support.TEXT}),
        ('one', None), ('one', {b'z': b'9007199254740993'}), ('one', {b'z': b''}),
        ('repeat', None), ('repeat', {b'z': b'copy'}),
        ('binary', {support.BINARY: b'valid', b'': b''}),
        ('binary', {support.BINARY: bytes(range(256)), b'': b''}),
        ('many', {f'f{i:03}'.encode(): f'v{128-i:03}'.encode() for i in range(0, 129, 2)}),
        ('within', {b'z': b'old', b'a': b'before'}),
        ('earlier', {b'z': b'old', b'a': b'before'}),
        ('head', {b'z': b'present'}),
        ('computed', {b'z': b'Z', b'a': b'A'}),
        ('lookup', b'wrong'), ('lookup', [b'm']), ('lookup', {b'm'}),
    ]
    rows = []
    for entry, initial in inputs:
        if initial is not None and not isinstance(initial, dict):
            rows.append((entry, initial, WRONG, initial, 'status'))
            continue
        expected = [(initial or {}).get(field) for field in QUERY['lookup' if entry == 'earlier' else entry]]
        if entry == 'head':
            rows.append((entry, initial, None, initial, 'null'))
        else:
            rows.append((entry, initial, expected, None if entry in ('within', 'earlier') else initial, 'array'))
    require(len(rows) == 18, 'HMGET case inventory')
    return rows


def source(seed=None):
    declarations = ['module HmgetCases', 'schema jobs : String -> Key Hash tag b"queue"',
        'def prepend : Bytes -> BulkArgs -> BulkArgs := fun (b : Bytes) (rest : BulkArgs) => bulkMore b rest',
        'def expose : Reply -> Reply := fun (r : Reply) => case r as x in Reply return Reply with',
        '| nil => nil | int n => int n | bulk b => bulk b',
        '| status b => status b | err b => status b | array rs => array rs',
        'def first : Replies -> Reply := fun (rs : Replies) => case rs as x in Replies return Reply with',
        '| repliesNil => nil | repliesCons r rest => r',
        'def headReply : Reply -> Reply := fun (r : Reply) => case r as x in Reply return Reply with',
        '| nil => nil | int n => int n | bulk b => bulk b',
        '| status b => status b | err b => status b | array rs => first rs',
        f'def removeScript : Script Reply {TAG} := do {{ r <- del Reply Hash {TAG} (jobs b"mail"); pure Reply {TAG} r }}']
    for entry, fields in QUERY.items():
        result = 'r' if entry == 'raw' else 'headReply r' if entry == 'head' else 'expose r'
        mutation = f'_ <- del Reply Hash {TAG} (jobs b"mail"); ' if entry == 'within' else ''
        cmd = command(fields)
        if entry == 'computed':
            cmd = cmd.replace(arguments(fields), f'(prepend {tet(fields[0])} {arguments(fields[1:])})')
        declarations += [f'def {entry}Script : Script Reply {TAG} := do {{ r <- {cmd}; '
            f'{mutation}pure Reply {TAG} ({result}) }}',
            f'def {entry} : Client Reply := do {{ r <- inv Reply {TAG} {entry}Script; done Reply r }}']
    declarations += [f'def earlier : Client Reply := do {{ r <- inv Reply {TAG} lookupScript; '
        f'_ <- inv Reply {TAG} removeScript; done Reply r }}']
    if seed is not None:
        entry, initial, *_ = cases()[seed]
        if isinstance(initial, dict):
            writes = [f'_ <- hset Reply {TAG} (jobs b"mail") {tet(f)} {tet(v)};' for f, v in initial.items()]
            declarations += [f'def seedScript : Script Reply {TAG} := do {{ ' + ' '.join(writes)
                + f' pure Reply {TAG} (status b"OK") }}',
                f'def seeded : Client Reply := do {{ _ <- inv Reply {TAG} seedScript; {entry} }}']
    return '\n'.join(declarations) + '\n'


def wire(expected, kind):
    if kind == 'array':
        return ('array:[' + ','.join('null' if v is None else 'bulk:' + v.hex() for v in expected) + ']').encode()
    return b'null' if kind == 'null' else kind.encode() + b':' + expected.hex().encode()


def stdout(expected, kind):
    if kind == 'array':
        values = [None if v is None else v.decode('utf-8') for v in expected]
        return json.dumps(values, ensure_ascii=False, separators=(',', ':')).replace('\x7f', '\\u007f').encode() + b'\n'
    return (expected or b'') + b'\n'


def emit(work, path, entry):
    output = work / ('out-' + entry)
    run(['./tether', 'emit', str(path), '--entry', entry, '-o', str(output)])
    for script in json.loads((output / 'scripts.json').read_text()):
        body = (output / (script['stem'] + '.lua')).read_bytes()
        require(body.startswith(b'#!lua flags=no-writes\n') == (script['entry'] in READONLY), 'HMGET write classification')
    return output


def twin(work, output, initial, expected, after, kind):
    value = wire(expected, kind) if kind == 'array' else (expected or b'')
    support.lua(work, output, initial, value, after, kind)


def offline(work, outputs):
    stores = twins = 0
    for i, (entry, initial, expected, after, kind) in enumerate(cases()):
        twin(work, outputs[entry], initial, expected, after, kind)
        twins += 1
        seed = (initial.decode() if isinstance(initial, bytes) else '@set' if isinstance(initial, set)
                else '@list' if isinstance(initial, list) else '@missing')
        (work / 'HmgetCases.tet').write_text(source(i))
        rows = []
        code = support.checker.check(work, 'HmgetCases.tet', command=[str(ROOT / '_build/default/dev/store_run.exe'),
            'HmgetCases.tet', 'seeded' if isinstance(initial, dict) else entry, KEY, seed, '1000000'], receive=rows.append)
        require(code == 0 and rows == [b'REPLY ' + wire(expected, kind) + b'\n'], f'HMGET store reply {i}: {rows!r}')
        stores += 1
    print(f'PASS HMGET-ORACLES store={stores} luajit={twins}', flush=True)


def refusals(work):
    cmd = command([b'a'])
    mismatch = b'CHECK mismatch: the constructor keyBytes of Key gives the index '
    invalid = [(cmd.replace(TAG, '(tag b"elsewhere")'), 'Hash', mismatch + b'(In SMu Tag [] (ACtor tag)')]
    invalid += [(cmd, schema, mismatch + f'(In SMu RedisType [] (ACtor {ctor})'.encode())
                for schema, ctor in [('(Str Binary)', 'Str'), ('Set', 'Set'), ('List', 'List'), ('ZSet', 'ZSet'), ('Stream', 'Stream')]]
    unbound = b'CHECK unbound: '
    invalid += [(cmd.replace(arguments([b'a']), bad), 'Hash', unbound + reason) for bad, reason in
                [('b"a"', b'bytesCons is not a constructor of BulkArgs'),
                 ('repliesNil', b'repliesNil is not a constructor of BulkArgs'),
                 ('(bulkOne nil)', b'nil is not a constructor of Bytes'),
                 ('(bulkMore b"a" repliesNil)', b'repliesNil is not a constructor of BulkArgs')]]
    for i, (body, schema, diagnostic) in enumerate(invalid):
        path, output = work / 'Refused.tet', work / f'refused-{i}'
        path.write_text(f'module Refused\nschema jobs : String -> Key {schema} tag b"queue"\n'
            f'def s : Script Reply {TAG} := do {{ r <- {body}; pure Reply {TAG} r }}\n'
            f'def main : Client Reply := do {{ r <- inv Reply {TAG} s; done Reply r }}\n')
        result = subprocess.run(['./tether', 'emit', str(path), '-o', str(output)], cwd=ROOT, capture_output=True, timeout=120)
        require(result.returncode == 2 and diagnostic in result.stderr, f'HMGET refusal {i}: {result.stderr!r}')
        require(not output.exists(), f'HMGET refusal published output {i}')
    print(f'PASS HMGET-REFUSALS cases={len(invalid)} atomic_output={len(invalid)}', flush=True)


def live(work, outputs):
    with support.local.hosts(work) as (port, env):
        redis = ['redis-cli', '-h', '127.0.0.1', '-p', str(port), '--raw']
        hosts = readonly = rejected = errors = 0
        live_cases = cases() + [('lookup', kind, WRONG, kind, 'status') for kind in ('zset', 'stream')]
        for entry, initial, expected, after, kind in live_cases:
            for host in ('node', 'bash'):
                run(redis + ['FLUSHDB'])
                run(redis + ['SET', 'other', 'kept'])
                if isinstance(initial, (set, list)):
                    operation = 'SADD' if isinstance(initial, set) else 'RPUSH'
                    for value in sorted(initial):
                        run(redis + ['EVAL', f"return redis.call('{operation}',KEYS[1]," + support.literal(value) + ')', '1', KEY])
                elif isinstance(initial, dict):
                    for field, value in reversed(sorted(initial.items())):
                        run(redis + ['EVAL', "return redis.call('HSET',KEYS[1]," + support.literal(field) + ',' + support.literal(value) + ')', '1', KEY])
                elif initial == 'zset':
                    run(redis + ['ZADD', KEY, '1', 'm'])
                elif initial == 'stream':
                    run(redis + ['XADD', KEY, '*', 'f', 'v'])
                elif initial is not None:
                    run(redis + ['-x', 'SET', KEY], data=initial)
                run(redis + ['PEXPIREAT', KEY, '4102444800000'])
                before, deadline = run(redis + ['DUMP', KEY]), run(redis + ['PEXPIRETIME', KEY])
                ro = entry not in ('within', 'earlier')
                run(redis + ['ACL', 'SETUSER', 'default', '-eval' if ro else '+eval', '-evalsha' if ro else '+evalsha'])
                output = outputs[entry]
                args = (['node', 'runtime/redis-host.mjs', str(output / 'prog.wasm'), str(port)]
                        if host == 'node' else ['/bin/bash', str(output / 'prog.sh')])
                try:
                    wanted, code = stdout(expected, kind), 0
                except UnicodeDecodeError:
                    wanted, code = b'', 4
                require(run(args, env=env, code=code) == wanted, f'HMGET {host} {entry} reply')
                run(redis + ['ACL', 'SETUSER', 'default', '+eval', '+evalsha'])
                if after == initial:
                    require(run(redis + ['DUMP', KEY]) == before and run(redis + ['PEXPIRETIME', KEY]) == deadline,
                            'HMGET preserves value and deadline')
                else:
                    require(run(redis + ['EXISTS', KEY]) == b'0\n', 'HMGET retained reply survives deletion')
                require(run(redis + ['GET', 'other']) == b'kept\n', 'HMGET unrelated key')
                hosts, readonly, rejected = hosts + 1, readonly + int(ro), rejected + int(code != 0)
        for host in ('node', 'bash'):
            run(redis + ['FLUSHDB'])
            run(redis + ['SET', KEY, 'wrong'])
            run(redis + ['ACL', 'SETUSER', 'default', '-eval', '-evalsha'])
            output = outputs['raw']
            args = (['node', 'runtime/redis-host.mjs', str(output / 'prog.wasm'), str(port)]
                    if host == 'node' else ['/bin/bash', str(output / 'prog.sh')])
            require(run(args, env=env, code=4) == b'', 'HMGET error stops host')
            require(run(redis + ['GET', KEY]) == b'wrong\n', 'HMGET error preserves value')
            run(redis + ['ACL', 'SETUSER', 'default', '+eval', '+evalsha'])
            hosts, readonly, errors = hosts + 1, readonly + 1, errors + 1
        require((hosts, readonly, rejected, errors) == (42, 38, 2, 2), 'HMGET host inventory')
        print(f'PASS HMGET-E2E cases={len(live_cases)} hosts={hosts} readonly={readonly} '
              f'utf8_refusals={rejected} errors={errors}', flush=True)
    count = 0
    for entry in ('main', 'retained'):
        for host in ('node', 'bash', 'luajit'):
            require(run(['./tether', 'exec', 'examples/ProfileFields.tet', '--entry', entry, '--host', host]) == EXAMPLE,
                    f'HMGET example {entry} {host}')
            count += 1
    print(f'PASS HMGET-EXAMPLE exec={count}', flush=True)


def main():
    probe = len(sys.argv) == 3 and sys.argv[1] == '--probe'
    require(probe or sys.argv[1:] in ([], ['--static'], ['--offline']), 'Usage: hmget-tests.py [--static|--offline|--probe ENTRY]')
    with tempfile.TemporaryDirectory(prefix='tether-hmget-') as temporary:
        work = Path(temporary)
        path = work / 'HmgetCases.tet'
        path.write_text(source())
        if probe:
            entry = sys.argv[2]
            rows = [row for row in cases() if row[0] == entry]
            require(entry in ENTRIES and bool(rows), 'HMGET probe must have cases')
            output = emit(work, path, entry)
            for _entry, initial, expected, after, kind in rows:
                twin(work, output, initial, expected, after, kind)
            print(f'PASS HMGET-PROBE entry={entry} cases={len(rows)}', flush=True)
            return
        outputs = {entry: emit(work, path, entry) for entry in ENTRIES}
        print(f'PASS HMGET-ARTIFACTS pairs={len(outputs)}', flush=True)
        refusals(work)
        for entry in ('main', 'retained'):
            require(run(['./tether', 'run', 'examples/ProfileFields.tet', '--entry', entry]) == EXAMPLE, 'HMGET interpreter example')
        if '--static' not in sys.argv:
            offline(work, outputs)
        if not sys.argv[1:]:
            live(work, outputs)
    print('PASS HMGET-TESTS' + (f' mode={sys.argv[1][2:]}' if sys.argv[1:] else ''), flush=True)


if __name__ == '__main__':
    try:
        main()
    except (AssertionError, OSError, ValueError, subprocess.SubprocessError) as error:
        print(f'FAIL HMGET-TESTS {error}', file=sys.stderr)
        sys.exit(1)

File: dev/hmget_tests.ml
module S = Tether_store.Store
module I = Tether_store.Interp
module E = Kanon_kernel.Eterm
let ( let* ) = Result.bind
let store result = Result.map_error S.message result
let require condition reason = if condition then Ok () else Error ("HMGET-UNIT " ^ reason)
let fold f xs = List.fold_left (fun acc x -> let* n = acc in let* () = f x in Ok (n + 1)) (Ok 0) xs
let rec term = function
  | I.Data (tid, tag, xs) -> E.KTag (tid, tag, List.map term xs)
  | I.Literal l -> E.KLit l
  | I.Fields _ | I.Closure _ | I.Erased -> E.KErased
let bytes s = term (I.bytes s)
let fields first rest =
  let rec make first = function
    | [] -> E.KTag (E.Tid "mu<BulkArgs>", 0, [bytes first])
    | next :: rest -> E.KTag (E.Tid "mu<BulkArgs>", 1, [bytes first; make next rest]) in
  make first rest
let execute ?(expose=false) args before =
  let key = E.KTag (E.Tid "mu<Key>", 0, [bytes "k"]) in
  let script = E.KTag (E.Tid "mu<Script>", 52, key :: args @ [E.KClos (E.Fid "pure", 1, [])]) in
  let client = E.KTag (E.Tid "mu<Client>", 1, [script; E.KClos (E.Fid "done", 1, [])]) in
  let fn name ps body = E.KFun (E.Fid name, ps, E.RI31, body) in
  let answer = if expose then E.KCase (E.Tid "mu<Reply>", E.KVar 0,
    [{E.tag=4; arity=1; body=E.KTag (E.Tid "mu<Reply>", 3, [E.KVar 0])}]) else E.KVar 0 in
  let rows = ["main", Kanon_kernel.Erase.Code [fn "main" [] client;
    fn "pure" [E.RI31] (E.KTag (E.Tid "mu<Script>", 0, [answer]));
    fn "done" [E.RI31] (E.KTag (E.Tid "mu<Client>", 0, [E.KVar 0]))]] in
  I.run ~budget:(Kanon_kernel.Budget.of_poll (fun () -> false)) rows ~entry:"main" before
let run () =
  let octets = List.of_seq (String.to_seq I.octets) |> List.map (String.make 1) in
  let cases = [
    None, "missing", [], [None];
    None, "a", ["b"; "a"], [None; None; None];
    Some [], "", [""], [None; None];
    Some ["z", "first"; "a", "second"], "z", ["a"; "z"], [Some "first"; Some "second"; Some "first"];
    Some ["a", "value"], "absent", ["a"; "absent"; "a"; "absent"], [None; Some "value"; None; Some "value"; None];
    Some ["", ""; "a", ""], "", ["a"; "no"], [Some ""; Some ""; None];
    Some [I.octets, I.octets], I.octets, ["no"; I.octets], [Some I.octets; None; Some I.octets];
    Some ["large", "9007199254740993"], "large", [], [Some "9007199254740993"];
    Some (List.map (fun b -> b, b) octets), "absent", octets, None :: List.map Option.some octets;
    Some ["a", "value"], "a", List.init 129 (fun _ -> "a"), List.init 130 (fun _ -> Some "value")]
  in
  let* values = fold (fun (initial, first, rest, expected) ->
    let before = S.put "other" (S.Str "kept") S.empty in
    let before = Option.fold ~none:before ~some:(fun fs -> S.put "k" (S.Hash fs) before) initial in
    let* _, before = store (S.expire ~seconds:false "k" "5000" before) in
    let* actual = store (S.hmget "k" first rest before) in
    let* () = require (actual = expected) "store preserves order, nils and duplicates" in
    let* reply, after = execute [fields first rest] before in
    require (reply = I.Array (List.map (Option.fold ~none:I.Nil ~some:(fun s -> I.Bulk s)) expected)
      && after = before) "interpreter reply and complete state") cases in
  let* errors = fold (fun value ->
    let before = S.put "k" value (S.put "other" (S.Str "kept") S.empty) in
    let* _, before = store (S.expire ~seconds:false "k" "5000" before) in
    let* () = require (S.hmget "k" "a" [] before = Error S.Wrong_type) "store wrong type" in
    let* () = require (execute [fields "a" []] before = Error (S.message S.Wrong_type)) "error stops client" in
    let* reply, after = execute ~expose:true [fields "a" []] before in
    require (reply = I.Status (S.message S.Wrong_type) && after = before) "error preserves complete state")
    [S.Str "old"; S.Set ["m"]; S.List ["m"]; S.ZSet ["m", "1"]; S.Stream []] in
  let bad tag xs = E.KTag (E.Tid "mu<BulkArgs>", tag, xs) in
  let* shapes = fold (fun (args, expected) ->
    require (execute args S.empty = Error expected) "malformed operand refusal") [
      [], "STORE-SCRIPT-COMMAND"; [bytes "a"], "STORE-SCRIPT-COMMAND";
      [fields "a" []; fields "b" []], "STORE-SCRIPT-COMMAND";
      [bad 0 []], "STORE-BULK-ARGS"; [bad 0 [bytes "a"; bytes "b"]], "STORE-BULK-ARGS";
      [bad 1 [bytes "a"]], "STORE-BULK-ARGS"; [bad 2 [bytes "a"]], "STORE-BULK-ARGS";
      [bad 1 [bytes "a"; bytes "b"]], "STORE-BULK-ARGS";
      [bad 0 [E.KErased]], "STORE-BYTES";
      [bad 1 [E.KErased; fields "a" []]], "STORE-BYTES"] in
  let before = S.put "k" (S.Hash ["a", "value"]) S.empty in
  let* _, expiring = store (S.expire ~seconds:false "k" "1" before) in
  let* expired = store (S.advance "2" expiring) in
  let* reply, after = execute [fields "a" ["a"]] expired in
  let* () = require (reply = I.Array [I.Nil; I.Nil] && after = expired) "expired Hash is missing" in
  let count = values + errors + shapes + 1 in
  let* () = require (count = 26) "case inventory" in Ok count
let () = run () |> Result.fold
  ~ok:(fun count -> Printf.printf "PASS HMGET-UNIT cases=%d\n" count)
  ~error:(fun e -> prerr_endline ("FAIL " ^ e); exit 1)
